@albanD's  #15219 augmented with SavedVariable::weak_grad_fn_

skrah opened the pull request:
Fixes #10532 via #15219 and lots of analysis by @albanD.

skrah committed:
Make SavedVariable::grad_fn_ a weak reference.

skrah committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into saved_variable_weakref

ezyang commented:
Better write a test :)

albanD commented:
What happens if the Function is only kept alive by the SavedVariable? It will be deleted and can never be recovered no?

apaszke commented:
Why is `shared_from_this()` necessary for inputs now? Also, is there a description of the reference cycle somewhere?

skrah commented:
> Better write a test :)

Yes, definitely, I'm still viewing this diff with a lot of suspicion myself. :)

I've been working on completely different things this week (xnd); I hope I can answer @albanD's and @apaszke's questions in more depth soon.

@apaszke, meanwhile the short answer is roughly this:

The reference cycle only occurs **after** applying @albanD's PR from #15219.   Unfortunately, Valgrind shows leaks at different places.  One such place is in `Variable::rebase_history()` where `copy_slices` leaks.

Another place is in `Variable::grad_fn()`, where `fn` leaks.

Here's a short reproducer:

```python
import torch
import gc

torch.set_num_threads(1)


# Test an in-place operation on a view in which the in-place op saves
# its output. Previously, this created a reference cycle.
dealloc = [0]

class IncrementOnDelete(object):
    def __del__(self):
        dealloc[0] += 1

c = IncrementOnDelete()

def test():
    root = torch.randn(3, 3, requires_grad=True)
    copy = root.clone()
    copy.grad_fn.register_hook(c)
    view = copy.view(9)
    torch.nn.functional.relu(view, inplace=True)

test()
c = None
gc.collect()
print(gc.garbage)
print(dealloc)
```

The output in `master` should be:

```
[]
[1]
```

The output after  #15219 only:

```
[]
[0]
```

The output after #15219 **and** #23502:

```
[]
[1]
```


Using `shared_fom_this()` also for inputs prevents https://github.com/pytorch/pytorch/blob/b2f6e2bdc1373e799f20adaa97ea287a06f03578/torch/csrc/autograd/saved_variable.cpp#L49 from being triggered by the test suite.

From memory, such failures were in `AbsBackward()` and other unary functions, but I may be wrong.

skrah committed:
Add SavedVariable::weak_grad_fn_ specifically for inplace views.

  - This is a far more conservative approach that does not have any effect
    on the current code base. It is a preparation for #15219.

albanD commented:
I would have a small question in the code sample below (I don't have a workstation so I can't easily compile and run your patch directly :/ )

```python
import torch

a = torch.rand(10, requires_grad=True)

b = a.narrow(0, 0, 2)
print(b.grad_fn)

b.add_(3)
print(b.grad_fn)
# What will keep `AsStridedBackward` alive after this patch here?
# Since b will only have a weak pointer to it.
```

skrah commented:
Before looking at @albanD's code sample, just a short note how the latest version of this PR is supposed to work.  It introduces an additional flag for in-place views, which is always `false` in `master`, so fortunately this version does nothing yet, which feels a lot safer. :)

#15219 would need to be modified slightly to use this flag, e.g.:

```
diff --git a/tools/autograd/gen_variable_type.py b/tools/autograd/gen_variable_type.py
index 5a02c93..74a8e9b 100644
--- a/tools/autograd/gen_variable_type.py
+++ b/tools/autograd/gen_variable_type.py
@@ -658,9 +658,13 @@ def emit_body(declaration):
                 if var == 'self' and inplace:
                     var = 'self.clone()'
                     assert not is_output
+                check_inplace_view = "false"
                 if inplace and is_output:
                     var = 'self'
-                expr = 'SavedVariable({}, {})'.format(var, str(is_output).lower())
+                    is_inplace_view = "as_variable_ref({}).is_view()".format(var)
+                    expr = 'SavedVariable({}, {}, {})'.format(var, str(is_output).lower(), is_inplace_view)
+                else:
+                    expr = 'SavedVariable({}, {})'.format(var, str(is_output).lower())
             elif arg['type'] == 'TensorList':
                 name += '_'
                 expr = 'make_saved_variable_list({})'.format(arg['name'])
```

With this modification all tests seem to pass and I don't see any leaks.  Perhaps there is a better location to break the cycle than `weak_grad_fn_` but I haven't found it yet. ;)

skrah commented:
@albanD I couldn't find a call to `SavedVariable()` in the sample, so I tried this:

```
import torch
import time
import gc

a = torch.rand(10, requires_grad=True)

b = a.narrow(0, 0, 2)
print(b.grad_fn)

b.relu_()
gc.collect(); print(b.grad_fn)
b.add_(3)
gc.collect(); print(b.grad_fn)
# What will keep `AsStridedBackward` alive after this patch here?
# Since b will only have a weak pointer to it.
```

This is in `relu_()`:

```
[VariableType_0.cpp]
9712      if (grad_fn) {
9713          rebase_history(flatten_tensor_args( self ), grad_fn);
```

This is the place where `b==self` gets a new `grad_fn_`:

```
[variable.cpp]
155           diff_view_meta->grad_fn_ = std::move(fn);
(gdb) p diff_view_meta->grad_fn_
$4 = std::shared_ptr (count 1, weak 1) 0x1dc3130
```

This is at the end of `relu_()`, where `SavedVariable()` is called:

```
[VariableType_0.cpp]
9720        grad_fn->result_ = SavedVariable(self, true, as_variable_ref(self).is_view());
(gdb) p as_variable_ref(self).grad_fn()
$8 = std::shared_ptr (count 1, weak 1) 0x1dc3130
```

So `b==self` owns `AsStridedBackward`, and `grad_fn->result_` gets the weak pointer.

Is there a scenario where `b` can disappear and `grad_fn->result_` is still used?

albanD commented:
> Is there a scenario where b can disappear and grad_fn->result_ is still used?

I think so as the graph never contain Variables, just a bunch of Nodes (previously named Functions) that contain the necessary SavedVariables.
If you do `c = b.clone()` then `del b` (or as users often do: `b = some_outofplace_op(b)`). The clone function will point directly to the AsStridedBackward and the `b` Variable will disappear. But if you backward in this graph, the AsStridedBackward.result_ will (potentially) be needed by the backward computation.
I may be mistaken here! Do let me know if you agree with this or not?

skrah commented:
Thanks, you're right that `b` can disappear.  However, in the following example,  the shared pointer is still alive:

```
import torch
import time
import gc

a = torch.rand(10, requires_grad=True)

b = a.narrow(0, 0, 2).clone().view(-1)
b.relu_()
gc.collect(); print(b.grad_fn)

c = b.clone()
del b
gc.collect(); print(c.grad_fn)

s = c.sum()
print("starting backward")
s.backward()
```

At the end of `relu_`, with `b==self`:

```
(gdb) p self
$55 = (at::Tensor &) @0xb6e158: {impl_ = {target_ = 0xb6c2d0}}
(gdb) p as_variable_ref(self).grad_fn()
$56 = std::shared_ptr (count 1, weak 2) 0x1dc82a0
```

Later, in `SharedVariable.unpack()`:

```
47        auto grad_fn = is_inplace_view_ ? weak_grad_fn_.lock() : grad_fn_;
(gdb) p grad_fn
$58 = std::shared_ptr (count 2, weak 2) 0x1dc82a0
(gdb) p *((at::Tensor *)0xb6e158)
$59 = {impl_ = {target_ = 0x2}}
(gdb) p is_inplace_view_
$61 = true
```

So something has preserved `grad_fn` while `b` is now undefined.

skrah commented:
> So something has preserved grad_fn while b is now undefined.

It appears that in`b.clone()` the `grad_fn` is preserved:

```
[torch::autograd::Variable::gradient_edge]
210         if (const auto& gradient = grad_fn()) {
211           return Edge(gradient, output_nr());
```

In line 211 the use count of the shared pointer increases to 2, so that's why `b` can be deleted safely  and we still have use count 1 in the end.

albanD reviewed (approved):
So after getting lost for a while I found the problem I remembered with this: if you do two inplace ops on a view one after the other where the first one saves some variables and you double backward into it 🎉 

The first part of the comment will explain that case. Let me know if you agree this case is problematic, the second part discuss if this case is useful or not

## First part 

Note in the pictures below the link from `a` to the AccumulateGrad is not necessarily correct but not relevant for the discussion.

In the first case (top part of the picture below), with
```python
a = torch.rand(10, requires_grad=True).narrow(0, 0, 10)

b = a.relu_()
```
Here, the ReluBackward will be kept alive either by `b` or the next Node that will be added (like clone in the example above).
So in this case, a weak pointer is fine as the AsStridedBackward is needed from the SavedVariable if and only if either `b` is still alive and we did backward on it or another Node has been created and it references the AsStridedBackward thus keeping everything alive.

In the second case (bottom part of the picture below)
```python
a = torch.rand(10, requires_grad=True).narrow(0, 0, 10)

b = a.relu_()

c = b.add_()
# That is the graph in the picture.

del b
# Now the AsStridedBackward goes away

c.sum().backward(torch.ones(1, requires_grad=True, create_graph=True))
# Here the double backward of the relu should point to the AsStridedBackward that disapeared.
# You can find more informations about these graphs in the forum discussion here https://discuss.pytorch.org/t/freeing-buffer-strange-behavior/31955/8
```

In the picture, the squares are Variables (or SavedVariables) and the circles are Nodes.
Hard line show strong references while dotted ones show weak references.
![IMG_20190806_161647](https://user-images.githubusercontent.com/6359743/62547836-277a1d00-b866-11e9-8d19-2430b275b7cb.jpg)


## Second part
So the problem only occurs when an inplace op happens on a view with another inplace op that saved some output).
One thing I did not thought about last time I looked at this:
But I'm not sure this is allowed: if we saved the Variable, that means that it is needed for the backward computation. But if we do another inplace op on it, then the backward of the first inplace op will fail as the saved variable has been modified !
In that case, this PR is fine and as the error for part 1 cannot happen as it will crash because of the modified variable before.

## Conclusion
If you agree with all above, then I would say this PR is good to go !
Do add a comment about the chaining of inplace ops please though for next time.

skrah commented:
[Second part]
> So the problem only occurs when an inplace op happens on a view with another inplace op that saved some output).

Empirically it seems okay, because torch raises an error here (both in `master` and with #23502 + #15219):

```
import torch

a = torch.rand(10, requires_grad=True).narrow(0, 0, 10)

b = a.relu_()

c = b.add_(100)
# That is the graph in the picture.

del b
# Now the AsStridedBackward goes away

c.sum().backward(torch.ones(1, requires_grad=True))
```

This gives:

```
RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation: [torch.FloatTensor [10]], which is output 0 of ReluBackward1, is at version 2; expected version 1 instead. Hint: enable anomaly detection to find the operation that failed to compute its gradient, with torch.autograd.set_detect_anomaly(True).
```

skrah committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into saved_variable_weakref

albanD commented:
@ezyang this is good to go on my side.

apaszke commented:
This weird interaction between `AsStridedBackward` and `CopySlices` makes me wonder if we shouldn't just collapse those two into a single op, and simplify the issue... It's not clear at all that it's either the case that _both_ or _none_ of them exist at the same time, and this is crucial for correctness of this patch.

ezyang commented:
I'm finally taking a look. Two high level comments:

1. Still needs tests
2. I think it would make sense to combine the two patches together, rather than trying to land them separately.

skrah commented:
@ezyang I think combining the patches is a good idea; probably the combined patch should be in #15219,  since this is more of a supporting PR.  @albanD, what do you think?

Then it is also possible to write tests.

I'm also still thinking about @apaszke's suggestion to merge `AsStridedBackward` and `CopySlices`, but on the surface that looks like a larger issue to me.

skrah committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into saved_variable_weakref

ezyang commented:
> I'm also still thinking about @apaszke's suggestion to merge AsStridedBackward and CopySlices, but on the surface that looks like a larger issue to me.

I'm not sure about it either. I have to think about the code.

ezyang reviewed (changes requested):
Please merge this with the other patch and add tests

skrah committed:
Merge @albanD's #15219 into this one.

skrah committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into saved_variable_weakref

skrah committed:
Add tests.

skrah committed:
Update comment.

skrah committed:
Fix flake8.

ezyang commented on torch/csrc/autograd/saved_variable.h:
```diff
@@ -52,5 +55,6 @@ class TORCH_API SavedVariable {
   bool was_default_constructed_ = true;
   bool requires_grad_ = false;
   bool has_grad_fn_ = false;
+  bool is_inplace_view_ = false;
```
We had a lot of good discussion in the PR, and it would be very helpful this discussion was replicated in a more durable form in the source code itself, in case someone is reading the code in question.

ezyang reviewed (approved):
After thinking about the examples in question, I think @apaskze's suggestion is reasonable, but this patch is so short that we should definitely land it, even if we want to do the bigger fix.
